### Problem Statement:

    By using Pandas and numpy library analyse the google play Store dataset and 
    Do the following task.

    1.Import the library
    2.Import the dataset
    3.Remove the Duplicate values
    4.Missing Value treatment
    5.Make data  ready for implementing ML Model.


## 1. Importing libraries
NumPy is the fundamental package for scientific computing with Python
Pandas is for data manipulation and analysis
Warning messages are typically issued in situations where it is useful to alert the user of some condition in a program, where that condition (normally) doesn’t warrant raising an exception and terminating the program.

In [71]:
import pandas as pd
import numpy as np
import warnings
warnings.filterwarnings('ignore')

## 2. Importing dataset
Dataset provided by Decodr technologies as part of an assignment 

In [431]:
gg=pd.read_csv("Googleplaystore .csv")

In [432]:
gg.shape # The dataset has 10841 observations and 13 features.

(10841, 13)

In [433]:
gg.columns # lising of columns.

Index(['App', 'Category', 'Rating', 'Reviews', 'Size', 'Installs', 'Type',
       'Price', 'Content Rating', 'Genres', 'Last Updated', 'Current Ver',
       'Android Ver'],
      dtype='object')

In [434]:
gg.info() # Basic information about the columns

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10841 entries, 0 to 10840
Data columns (total 13 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   App             10841 non-null  object 
 1   Category        10841 non-null  object 
 2   Rating          9367 non-null   float64
 3   Reviews         10841 non-null  object 
 4   Size            10841 non-null  object 
 5   Installs        10841 non-null  object 
 6   Type            10840 non-null  object 
 7   Price           10841 non-null  object 
 8   Content Rating  10840 non-null  object 
 9   Genres          10841 non-null  object 
 10  Last Updated    10841 non-null  object 
 11  Current Ver     10833 non-null  object 
 12  Android Ver     10838 non-null  object 
dtypes: float64(1), object(12)
memory usage: 1.1+ MB


In [435]:
gg.describe(include='all').T # /statistical information of all features.

count unique                 top   freq      mean       std  \
App              10841   9660              ROBLOX      9       NaN       NaN   
Category         10841     34              FAMILY   1972       NaN       NaN   
Rating          9367.0    NaN                 NaN    NaN  4.193338  0.537431   
Reviews          10841   6002                   0    596       NaN       NaN   
Size             10841    462  Varies with device   1695       NaN       NaN   
Installs         10841     22          1,000,000+   1579       NaN       NaN   
Type             10840      3                Free  10039       NaN       NaN   
Price            10841     93                   0  10040       NaN       NaN   
Content Rating   10840      6            Everyone   8714       NaN       NaN   
Genres           10841    120               Tools    842       NaN       NaN   
Last Updated     10841   1378      August 3, 2018    326       NaN       NaN   
Current Ver      10833   2832  Varies with device   1459       NaN       NaN   
Android Ver      10838     33          4.1 and up   2451       NaN       NaN   

                min  25%  50%  75%   max  
App             NaN  NaN  NaN  NaN   NaN  
Category        NaN  NaN  NaN  NaN   NaN  
Rating          1.0  4.0  4.3  4.5  19.0  
Reviews         NaN  NaN  NaN  NaN   NaN  
Size            NaN  NaN  NaN  NaN   NaN  
Installs        NaN  NaN  NaN  NaN   NaN  
Type            NaN  NaN  NaN  NaN   NaN  
Price           NaN  NaN  NaN  NaN   NaN  
Content Rating  NaN  NaN  NaN  NaN   NaN  
Genres          NaN  NaN  NaN  NaN   NaN  
Last Updated    NaN  NaN  NaN  NaN   NaN  
Current Ver     NaN  NaN  NaN  NaN   NaN  
Android Ver     NaN  NaN  NaN  NaN   NaN

#### After importing the dataset, you can see we use head function ( This function returns the first n rows for the object based on position. It is useful for quickly testing if your object has the right type of data in it. By default it returns 5 rows. )

In [436]:
gg.head() # listing of first five columns

App        Category  Rating  \
0     Photo Editor & Candy Camera & Grid & ScrapBook  ART_AND_DESIGN     4.1   
1                                Coloring book moana  ART_AND_DESIGN     3.9   
2  U Launcher Lite – FREE Live Cool Themes, Hide ...  ART_AND_DESIGN     4.7   
3                              Sketch - Draw & Paint  ART_AND_DESIGN     4.5   
4              Pixel Draw - Number Art Coloring Book  ART_AND_DESIGN     4.3   

  Reviews  Size     Installs  Type Price Content Rating  \
0     159   19M      10,000+  Free     0       Everyone   
1     967   14M     500,000+  Free     0       Everyone   
2   87510  8.7M   5,000,000+  Free     0       Everyone   
3  215644   25M  50,000,000+  Free     0           Teen   
4     967  2.8M     100,000+  Free     0       Everyone   

                      Genres      Last Updated         Current Ver  \
0               Art & Design   January 7, 2018               1.0.0   
1  Art & Design;Pretend Play  January 15, 2018               2.0.0   
2               Art & Design    August 1, 2018               1.2.4   
3               Art & Design      June 8, 2018  Varies with device   
4    Art & Design;Creativity     June 20, 2018                 1.1   

    Android Ver  
0  4.0.3 and up  
1  4.0.3 and up  
2  4.0.3 and up  
3    4.2 and up  
4    4.4 and up

#### We can combine tables to make it easier to see data. For this, we can use "concat function" that is found Pandas library.

pd.concat([data frame parameters], axis,ignore_index) : It combines 2 tables.
axis : It adds the tables as horizontal or vertical. If axis equals 0, it adds as horizontal. If axis equals 1, it adds as vertical.
ignore_index : It ignores index values.

In [437]:
gg1 = gg.head()
gg2 = gg.tail()
gg_full = pd.concat([gg1,gg2],axis=0,ignore_index=True)
gg_full

App             Category  \
0     Photo Editor & Candy Camera & Grid & ScrapBook       ART_AND_DESIGN   
1                                Coloring book moana       ART_AND_DESIGN   
2  U Launcher Lite – FREE Live Cool Themes, Hide ...       ART_AND_DESIGN   
3                              Sketch - Draw & Paint       ART_AND_DESIGN   
4              Pixel Draw - Number Art Coloring Book       ART_AND_DESIGN   
5                                   Sya9a Maroc - FR               FAMILY   
6                   Fr. Mike Schmitz Audio Teachings               FAMILY   
7                             Parkinson Exercices FR              MEDICAL   
8                      The SCP Foundation DB fr nn5n  BOOKS_AND_REFERENCE   
9      iHoroscope - 2018 Daily Horoscope & Astrology            LIFESTYLE   

   Rating Reviews                Size     Installs  Type Price Content Rating  \
0     4.1     159                 19M      10,000+  Free     0       Everyone   
1     3.9     967                 14M     500,000+  Free     0       Everyone   
2     4.7   87510                8.7M   5,000,000+  Free     0       Everyone   
3     4.5  215644                 25M  50,000,000+  Free     0           Teen   
4     4.3     967                2.8M     100,000+  Free     0       Everyone   
5     4.5      38                 53M       5,000+  Free     0       Everyone   
6     5.0       4                3.6M         100+  Free     0       Everyone   
7     NaN       3                9.5M       1,000+  Free     0       Everyone   
8     4.5     114  Varies with device       1,000+  Free     0     Mature 17+   
9     4.5  398307                 19M  10,000,000+  Free     0       Everyone   

                      Genres      Last Updated         Current Ver  \
0               Art & Design   January 7, 2018               1.0.0   
1  Art & Design;Pretend Play  January 15, 2018               2.0.0   
2               Art & Design    August 1, 2018               1.2.4   
3               Art & Design      June 8, 2018  Varies with device   
4    Art & Design;Creativity     June 20, 2018                 1.1   
5                  Education     July 25, 2017                1.48   
6                  Education      July 6, 2018                 1.0   
7                    Medical  January 20, 2017                 1.0   
8          Books & Reference  January 19, 2015  Varies with device   
9                  Lifestyle     July 25, 2018  Varies with device   

          Android Ver  
0        4.0.3 and up  
1        4.0.3 and up  
2        4.0.3 and up  
3          4.2 and up  
4          4.4 and up  
5          4.1 and up  
6          4.1 and up  
7          2.2 and up  
8  Varies with device  
9  Varies with device

### 1. Category- Study of feature one.
     # This feature has 34 unique features of which one feature 1.9 is not in concurrence with other features. The correponding row was identified and dropped from the dataset.

In [438]:
gg.Category.nunique()

34

In [439]:
gg.Category.unique()

array(['ART_AND_DESIGN', 'AUTO_AND_VEHICLES', 'BEAUTY',
       'BOOKS_AND_REFERENCE', 'BUSINESS', 'COMICS', 'COMMUNICATION',
       'DATING', 'EDUCATION', 'ENTERTAINMENT', 'EVENTS', 'FINANCE',
       'FOOD_AND_DRINK', 'HEALTH_AND_FITNESS', 'HOUSE_AND_HOME',
       'LIBRARIES_AND_DEMO', 'LIFESTYLE', 'GAME', 'FAMILY', 'MEDICAL',
       'SOCIAL', 'SHOPPING', 'PHOTOGRAPHY', 'SPORTS', 'TRAVEL_AND_LOCAL',
       'TOOLS', 'PERSONALIZATION', 'PRODUCTIVITY', 'PARENTING', 'WEATHER',
       'VIDEO_PLAYERS', 'NEWS_AND_MAGAZINES', 'MAPS_AND_NAVIGATION',
       '1.9'], dtype=object)

In [440]:
gg.Category.value_counts()

FAMILY                 1972
GAME                   1144
TOOLS                   843
MEDICAL                 463
BUSINESS                460
PRODUCTIVITY            424
PERSONALIZATION         392
COMMUNICATION           387
SPORTS                  384
LIFESTYLE               382
FINANCE                 366
HEALTH_AND_FITNESS      341
PHOTOGRAPHY             335
SOCIAL                  295
NEWS_AND_MAGAZINES      283
SHOPPING                260
TRAVEL_AND_LOCAL        258
DATING                  234
BOOKS_AND_REFERENCE     231
VIDEO_PLAYERS           175
EDUCATION               156
ENTERTAINMENT           149
MAPS_AND_NAVIGATION     137
FOOD_AND_DRINK          127
HOUSE_AND_HOME           88
LIBRARIES_AND_DEMO       85
AUTO_AND_VEHICLES        85
WEATHER                  82
ART_AND_DESIGN           65
EVENTS                   64
COMICS                   60
PARENTING                60
BEAUTY                   53
1.9                       1
Name: Category, dtype: int64

In [441]:
gg[gg['Category'] == '1.9']

App Category  Rating Reviews  \
10472  Life Made WI-Fi Touchscreen Photo Frame      1.9    19.0    3.0M   

         Size Installs Type     Price Content Rating             Genres  \
10472  1,000+     Free    0  Everyone            NaN  February 11, 2018   

      Last Updated Current Ver Android Ver  
10472       1.0.19  4.0 and up         NaN

In [442]:
gg.loc[10472] = gg.loc[10472].shift()
gg['App'].loc[10472] = gg['Category'].loc[10472]
gg['Category'].loc[10472] = np.nan
gg.loc[10472]

App               Life Made WI-Fi Touchscreen Photo Frame
Category                                              NaN
Rating                                                1.9
Reviews                                              19.0
Size                                                 3.0M
Installs                                           1,000+
Type                                                 Free
Price                                                   0
Content Rating                                   Everyone
Genres                                                NaN
Last Updated                            February 11, 2018
Current Ver                                        1.0.19
Android Ver                                    4.0 and up
Name: 10472, dtype: object

In [443]:
gg.drop(gg.index[10472], inplace=True)

In [444]:
gg.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 10840 entries, 0 to 10840
Data columns (total 13 columns):
 #   Column          Non-Null Count  Dtype 
---  ------          --------------  ----- 
 0   App             10840 non-null  object
 1   Category        10840 non-null  object
 2   Rating          9366 non-null   object
 3   Reviews         10840 non-null  object
 4   Size            10840 non-null  object
 5   Installs        10840 non-null  object
 6   Type            10839 non-null  object
 7   Price           10840 non-null  object
 8   Content Rating  10840 non-null  object
 9   Genres          10840 non-null  object
 10  Last Updated    10840 non-null  object
 11  Current Ver     10832 non-null  object
 12  Android Ver     10838 non-null  object
dtypes: object(13)
memory usage: 1.2+ MB


## 3. Remove duplicate values
    ##### If the duplicate or missing values are not handled properly by the researcher, then he/she may end up drawing an inaccurate inference about the data. Due to improper handling, the result obtained by the researcher will differ from ones where the missing values are present.

In [445]:
gg.duplicated().sum()

483

In [446]:
gg.drop_duplicates(keep='first',inplace=True)

In [447]:
gg.shape

(10357, 13)

In [448]:
gg.describe(include='all').T

count unique                 top    freq
App              10357   9659              ROBLOX       9
Category         10357     33              FAMILY    1943
Rating          8892.0   39.0                 4.4  1032.0
Reviews          10357   6001                   0     593
Size             10357    461  Varies with device    1526
Installs         10357     21          1,000,000+    1488
Type             10356      2                Free    9591
Price            10357     92                   0    9592
Content Rating   10357      6            Everyone    8382
Genres           10357    119               Tools     842
Last Updated     10357   1377      August 3, 2018     298
Current Ver      10349   2831  Varies with device    1302
Android Ver      10355     33          4.1 and up    2379

## 4 . Missing value treatment
    # After deleting duplicated values, it is noted that  there are 4 features with null values

In [449]:
gg.isnull().sum().sort_values(ascending=False)

Rating            1465
Current Ver          8
Android Ver          2
Type                 1
App                  0
Category             0
Reviews              0
Size                 0
Installs             0
Price                0
Content Rating       0
Genres               0
Last Updated         0
dtype: int64

### 2, Rating Feature 2
    #On a feature which has numeric data like rating column. We can calculate the mean, median or mode of the feature and replace it with the missing values.This is an approximation which can add variance to the data set. But the loss of the data can be negated by this method which yields better results compared to removal of rows and columns.Replacing with the above three approximations are a statistical approach of handling the missing values. 

In [450]:
gg['Rating'].mean()

4.187876743139908

In [451]:
gg['Rating'].median()

4.3

In [452]:
gg['Rating'].mode()

0    4.4
dtype: object

In [453]:
gg['Rating'].std()

0.5223767001082028

In [454]:
gg['Rating']=gg['Rating'].replace(np.nan,gg['Rating'].mean())
gg['Rating']

0        4.100000
1        3.900000
2        4.700000
3        4.500000
4        4.300000
           ...   
10836    4.500000
10837    5.000000
10838    4.187877
10839    4.500000
10840    4.500000
Name: Rating, Length: 10357, dtype: float64

In [455]:
gg.Rating.isna().sum()

0

In [456]:
gg.isnull().sum().sort_values(ascending=False)

Current Ver       8
Android Ver       2
Type              1
App               0
Category          0
Rating            0
Reviews           0
Size              0
Installs          0
Price             0
Content Rating    0
Genres            0
Last Updated      0
dtype: int64

In [457]:
gg['Rating'].tail()

10836    4.500000
10837    5.000000
10838    4.187877
10839    4.500000
10840    4.500000
Name: Rating, dtype: float64

In [458]:
gg.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 10357 entries, 0 to 10840
Data columns (total 13 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   App             10357 non-null  object 
 1   Category        10357 non-null  object 
 2   Rating          10357 non-null  float64
 3   Reviews         10357 non-null  object 
 4   Size            10357 non-null  object 
 5   Installs        10357 non-null  object 
 6   Type            10356 non-null  object 
 7   Price           10357 non-null  object 
 8   Content Rating  10357 non-null  object 
 9   Genres          10357 non-null  object 
 10  Last Updated    10357 non-null  object 
 11  Current Ver     10349 non-null  object 
 12  Android Ver     10355 non-null  object 
dtypes: float64(1), object(12)
memory usage: 1.1+ MB


In [459]:
gg.dtypes

App                object
Category           object
Rating            float64
Reviews            object
Size               object
Installs           object
Type               object
Price              object
Content Rating     object
Genres             object
Last Updated       object
Current Ver        object
Android Ver        object
dtype: object

In [460]:
gg.Rating.unique()

array([4.1       , 3.9       , 4.7       , 4.5       , 4.3       ,
       4.4       , 3.8       , 4.2       , 4.6       , 3.2       ,
       4.        , 4.18787674, 4.8       , 4.9       , 3.6       ,
       3.7       , 3.3       , 3.4       , 3.5       , 3.1       ,
       5.        , 2.6       , 3.        , 1.9       , 2.5       ,
       2.8       , 2.7       , 1.        , 2.9       , 2.3       ,
       2.2       , 1.7       , 2.        , 1.8       , 2.4       ,
       1.6       , 2.1       , 1.4       , 1.5       , 1.2       ])

### 3.  Type 
    # This feature has one null value which is located to index 9148  and it was chosen to drop since information the paid or free status for the same was not available.

In [461]:
gg['Type'].unique()

array(['Free', 'Paid', nan], dtype=object)

In [462]:
gg.index[gg['Type'].isna()]

Int64Index([9148], dtype='int64')

In [463]:
gg[gg['Type'].isnull()].index.tolist()

[9148]

In [464]:
gg.loc[9148]['Type']

nan

In [465]:
gg.loc[9148]

App               Command & Conquer: Rivals
Category                             FAMILY
Rating                             4.187877
Reviews                                   0
Size                     Varies with device
Installs                                  0
Type                                    NaN
Price                                     0
Content Rating                 Everyone 10+
Genres                             Strategy
Last Updated                  June 28, 2018
Current Ver              Varies with device
Android Ver              Varies with device
Name: 9148, dtype: object

In [466]:
gg.drop(axis=0, index=9148, inplace=True, errors='raise')

In [467]:
gg.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 10356 entries, 0 to 10840
Data columns (total 13 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   App             10356 non-null  object 
 1   Category        10356 non-null  object 
 2   Rating          10356 non-null  float64
 3   Reviews         10356 non-null  object 
 4   Size            10356 non-null  object 
 5   Installs        10356 non-null  object 
 6   Type            10356 non-null  object 
 7   Price           10356 non-null  object 
 8   Content Rating  10356 non-null  object 
 9   Genres          10356 non-null  object 
 10  Last Updated    10356 non-null  object 
 11  Current Ver     10348 non-null  object 
 12  Android Ver     10354 non-null  object 
dtypes: float64(1), object(12)
memory usage: 1.1+ MB


In [468]:
gg.isnull().sum().sort_values(ascending=False)

Current Ver       8
Android Ver       2
App               0
Category          0
Rating            0
Reviews           0
Size              0
Installs          0
Type              0
Price             0
Content Rating    0
Genres            0
Last Updated      0
dtype: int64

### 4. Content rating
    ## This feature has 2 observations with Unrated values and was chosen to be dropped, since it would affect the rating analysis

In [469]:
gg['Content Rating'].unique()

array(['Everyone', 'Teen', 'Everyone 10+', 'Mature 17+',
       'Adults only 18+', 'Unrated'], dtype=object)

In [470]:
gg.index[gg['Content Rating'] == 'Unrated']

Int64Index([7312, 8266], dtype='int64')

In [471]:
gg.drop(gg.index[[7312,8266]], inplace=True)

In [472]:
gg.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 10354 entries, 0 to 10840
Data columns (total 13 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   App             10354 non-null  object 
 1   Category        10354 non-null  object 
 2   Rating          10354 non-null  float64
 3   Reviews         10354 non-null  object 
 4   Size            10354 non-null  object 
 5   Installs        10354 non-null  object 
 6   Type            10354 non-null  object 
 7   Price           10354 non-null  object 
 8   Content Rating  10354 non-null  object 
 9   Genres          10354 non-null  object 
 10  Last Updated    10354 non-null  object 
 11  Current Ver     10346 non-null  object 
 12  Android Ver     10352 non-null  object 
dtypes: float64(1), object(12)
memory usage: 1.1+ MB


In [473]:
gg.isna().sum()

App               0
Category          0
Rating            0
Reviews           0
Size              0
Installs          0
Type              0
Price             0
Content Rating    0
Genres            0
Last Updated      0
Current Ver       8
Android Ver       2
dtype: int64

In [474]:
gg.loc[4453]

App               [substratum] Vacuum: P
Category                 PERSONALIZATION
Rating                               4.4
Reviews                              230
Size                                 11M
Installs                          1,000+
Type                                Paid
Price                              $1.49
Content Rating                  Everyone
Genres                   Personalization
Last Updated               July 20, 2018
Current Ver                          4.4
Android Ver                          NaN
Name: 4453, dtype: object

In [475]:
gg.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 10354 entries, 0 to 10840
Data columns (total 13 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   App             10354 non-null  object 
 1   Category        10354 non-null  object 
 2   Rating          10354 non-null  float64
 3   Reviews         10354 non-null  object 
 4   Size            10354 non-null  object 
 5   Installs        10354 non-null  object 
 6   Type            10354 non-null  object 
 7   Price           10354 non-null  object 
 8   Content Rating  10354 non-null  object 
 9   Genres          10354 non-null  object 
 10  Last Updated    10354 non-null  object 
 11  Current Ver     10346 non-null  object 
 12  Android Ver     10352 non-null  object 
dtypes: float64(1), object(12)
memory usage: 1.4+ MB


#### 5. Current version and Android Version
The null values in the Current Version and Android version columns were identified and was not taken up for the cleaning exercise

In [476]:
gg['Current Ver'].unique()

array(['1.0.0', '2.0.0', '1.2.4', ..., '1.0.612928', '0.3.4', '2.0.148.0'],
      dtype=object)

In [477]:
gg[gg['Current Ver'].isnull()].index.tolist()

[15, 1553, 6322, 6803, 7333, 7407, 7730, 10342]

In [478]:
gg['Android Ver'].unique()

array(['4.0.3 and up', '4.2 and up', '4.4 and up', '2.3 and up',
       '3.0 and up', '4.1 and up', '4.0 and up', '2.3.3 and up',
       'Varies with device', '2.2 and up', '5.0 and up', '6.0 and up',
       '1.6 and up', '1.5 and up', '2.1 and up', '7.0 and up',
       '5.1 and up', '4.3 and up', '4.0.3 - 7.1.1', '2.0 and up',
       '3.2 and up', '4.4W and up', '7.1 and up', '7.0 - 7.1.1',
       '8.0 and up', '5.0 - 8.0', '3.1 and up', '2.0.1 and up',
       '4.1 - 7.1.1', nan, '5.0 - 6.0', '1.0 and up', '2.2 - 7.1.1',
       '5.0 - 7.1.1'], dtype=object)

In [479]:
gg[gg['Android Ver'].isnull()].index.tolist()

[4453, 4490]

### 6 Reviews
    Incongruent values were replace and the feature converted to float type

In [480]:
gg.Reviews.unique()

array(['159', '967', '87510', ..., '603', '1195', '398307'], dtype=object)

In [481]:
gg['Reviews'] = gg.Reviews.replace("0.0",0)
gg['Reviews'] = gg.Reviews.replace("3.0M",3000000.0)
gg['Reviews'] = gg['Reviews'].astype(float)
gg['Reviews'].dtype

dtype('float64')

In [482]:
gg.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 10354 entries, 0 to 10840
Data columns (total 13 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   App             10354 non-null  object 
 1   Category        10354 non-null  object 
 2   Rating          10354 non-null  float64
 3   Reviews         10354 non-null  float64
 4   Size            10354 non-null  object 
 5   Installs        10354 non-null  object 
 6   Type            10354 non-null  object 
 7   Price           10354 non-null  object 
 8   Content Rating  10354 non-null  object 
 9   Genres          10354 non-null  object 
 10  Last Updated    10354 non-null  object 
 11  Current Ver     10346 non-null  object 
 12  Android Ver     10352 non-null  object 
dtypes: float64(2), object(11)
memory usage: 1.4+ MB


### 7 Size feature 
    All size values converted to  kilobyte type
    Values with Varies with device were alloted np.nan values since it was aimed to convert it to float datatype

In [483]:
gg['Size'].unique()

array(['19M', '14M', '8.7M', '25M', '2.8M', '5.6M', '29M', '33M', '3.1M',
       '28M', '12M', '20M', '21M', '37M', '2.7M', '5.5M', '17M', '39M',
       '31M', '4.2M', '7.0M', '23M', '6.0M', '6.1M', '4.6M', '9.2M',
       '5.2M', '11M', '24M', 'Varies with device', '9.4M', '15M', '10M',
       '1.2M', '26M', '8.0M', '7.9M', '56M', '57M', '35M', '54M', '201k',
       '3.6M', '5.7M', '8.6M', '2.4M', '27M', '2.5M', '16M', '3.4M',
       '8.9M', '3.9M', '2.9M', '38M', '32M', '5.4M', '18M', '1.1M',
       '2.2M', '4.5M', '9.8M', '52M', '9.0M', '6.7M', '30M', '2.6M',
       '7.1M', '3.7M', '22M', '7.4M', '6.4M', '3.2M', '8.2M', '9.9M',
       '4.9M', '9.5M', '5.0M', '5.9M', '13M', '73M', '6.8M', '3.5M',
       '4.0M', '2.3M', '7.2M', '2.1M', '42M', '7.3M', '9.1M', '55M',
       '23k', '6.5M', '1.5M', '7.5M', '51M', '41M', '48M', '8.5M', '46M',
       '8.3M', '4.3M', '4.7M', '3.3M', '40M', '7.8M', '8.8M', '6.6M',
       '5.1M', '61M', '66M', '79k', '8.4M', '118k', '44M', '695k', '1.6M',
     

In [484]:
gg['Size'] = gg.Size.replace("Varies with device",np.nan)
gg['Size'] = gg.Size.str.replace("M","000") # All size values became the kilobyte type.
gg['Size'] = gg.Size.str.replace("k","")
gg['Size'] = gg.Size.replace("1,000+",1000)
gg['Size'] =gg['Size'].astype(float)
gg['Size'].dtype

dtype('float64')

In [486]:
gg.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 10354 entries, 0 to 10840
Data columns (total 13 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   App             10354 non-null  object 
 1   Category        10354 non-null  object 
 2   Rating          10354 non-null  float64
 3   Reviews         10354 non-null  float64
 4   Size            8829 non-null   float64
 5   Installs        10354 non-null  object 
 6   Type            10354 non-null  object 
 7   Price           10354 non-null  object 
 8   Content Rating  10354 non-null  object 
 9   Genres          10354 non-null  object 
 10  Last Updated    10354 non-null  object 
 11  Current Ver     10346 non-null  object 
 12  Android Ver     10352 non-null  object 
dtypes: float64(3), object(10)
memory usage: 1.4+ MB


### We could also retain  Size to object type without converting  astype (float) but it may not lend itself to further data manipulation

### 8. Installs
    The values were altered to lend itself to conversion to float and so values like,+ etc were replaced with blank 

In [487]:
gg['Installs'].unique()

array(['10,000+', '500,000+', '5,000,000+', '50,000,000+', '100,000+',
       '50,000+', '1,000,000+', '10,000,000+', '5,000+', '100,000,000+',
       '1,000,000,000+', '1,000+', '500,000,000+', '50+', '100+', '500+',
       '10+', '1+', '5+', '0+'], dtype=object)

In [488]:
gg['Installs'] = gg.Installs.str.replace(",","")
gg['Installs'] = gg.Installs.str.replace("+","")
gg['Installs'] = gg.Installs.replace("Free",np.nan)
gg['Installs'] = gg['Installs'].astype(float)
gg['Installs'].dtype

dtype('float64')

In [489]:
gg.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 10354 entries, 0 to 10840
Data columns (total 13 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   App             10354 non-null  object 
 1   Category        10354 non-null  object 
 2   Rating          10354 non-null  float64
 3   Reviews         10354 non-null  float64
 4   Size            8829 non-null   float64
 5   Installs        10354 non-null  float64
 6   Type            10354 non-null  object 
 7   Price           10354 non-null  object 
 8   Content Rating  10354 non-null  object 
 9   Genres          10354 non-null  object 
 10  Last Updated    10354 non-null  object 
 11  Current Ver     10346 non-null  object 
 12  Android Ver     10352 non-null  object 
dtypes: float64(4), object(9)
memory usage: 1.4+ MB


###  9 PRICE
       #The dollar symbol wa replaced & converted to float type.

In [418]:
gg['Price'].unique()

array(['0', '$4.99', '$3.99', '$6.99', '$1.49', '$2.99', '$7.99', '$5.99',
       '$3.49', '$1.99', '$9.99', '$7.49', '$0.99', '$9.00', '$5.49',
       '$10.00', '$24.99', '$11.99', '$79.99', '$16.99', '$14.99',
       '$1.00', '$29.99', '$12.99', '$2.49', '$10.99', '$1.50', '$19.99',
       '$15.99', '$33.99', '$74.99', '$39.99', '$3.95', '$4.49', '$1.70',
       '$8.99', '$2.00', '$3.88', '$25.99', '$399.99', '$17.99',
       '$400.00', '$3.02', '$1.76', '$4.84', '$4.77', '$1.61', '$2.50',
       '$1.59', '$6.49', '$1.29', '$5.00', '$13.99', '$299.99', '$379.99',
       '$37.99', '$18.99', '$389.99', '$19.90', '$8.49', '$1.75',
       '$14.00', '$4.85', '$46.99', '$109.99', '$154.99', '$3.08',
       '$2.59', '$4.80', '$1.96', '$19.40', '$3.90', '$4.59', '$15.46',
       '$3.04', '$4.29', '$2.60', '$3.28', '$4.60', '$28.99', '$2.95',
       '$2.90', '$1.97', '$200.00', '$89.99', '$2.56', '$30.99', '$3.61',
       '$394.99', '$1.26', '$1.20', '$1.04'], dtype=object)

In [490]:
gg['Price'] = gg.Price.str.replace("$","")
gg['Price'] = gg['Price'].astype(float)
gg['Price'].dtype

dtype('float64')

In [491]:
gg.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 10354 entries, 0 to 10840
Data columns (total 13 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   App             10354 non-null  object 
 1   Category        10354 non-null  object 
 2   Rating          10354 non-null  float64
 3   Reviews         10354 non-null  float64
 4   Size            8829 non-null   float64
 5   Installs        10354 non-null  float64
 6   Type            10354 non-null  object 
 7   Price           10354 non-null  float64
 8   Content Rating  10354 non-null  object 
 9   Genres          10354 non-null  object 
 10  Last Updated    10354 non-null  object 
 11  Current Ver     10346 non-null  object 
 12  Android Ver     10352 non-null  object 
dtypes: float64(5), object(8)
memory usage: 1.4+ MB


### 10. Last Updated
    # Pandas to_datetime function was used to  convert object type to datetime64 format.

In [421]:
gg['Last Updated'].unique()

array(['January 7, 2018', 'January 15, 2018', 'August 1, 2018', ...,
       'January 20, 2014', 'February 16, 2014', 'March 23, 2014'],
      dtype=object)

In [422]:
gg['Last Updated'] = pd.to_datetime(gg['Last Updated'])
gg['Last Updated']

0       2018-01-07
1       2018-01-15
2       2018-08-01
3       2018-06-08
4       2018-06-20
           ...    
10836   2017-07-25
10837   2018-07-06
10838   2017-01-20
10839   2015-01-19
10840   2018-07-25
Name: Last Updated, Length: 10354, dtype: datetime64[ns]

In [423]:
gg.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 10354 entries, 0 to 10840
Data columns (total 13 columns):
 #   Column          Non-Null Count  Dtype         
---  ------          --------------  -----         
 0   App             10354 non-null  object        
 1   Category        10354 non-null  object        
 2   Rating          10354 non-null  float64       
 3   Reviews         10354 non-null  float64       
 4   Size            8829 non-null   float64       
 5   Installs        10354 non-null  float64       
 6   Type            10354 non-null  object        
 7   Price           10354 non-null  float64       
 8   Content Rating  10354 non-null  object        
 9   Genres          10354 non-null  object        
 10  Last Updated    10354 non-null  datetime64[ns]
 11  Current Ver     10346 non-null  object        
 12  Android Ver     10352 non-null  object        
dtypes: datetime64[ns](1), float64(5), object(7)
memory usage: 1.4+ MB


### Finally we are left with null values in Size corresponding to the Varies with device values in the original dataset. and null values in Current version & Android version which can for now be ignored.

In [ ]:
gg.isnull().sum().sort_values(ascending=False)

### Original data had 10841 rows of which  1 category observation , 483 duplicated values,2 unrated content rating and one null value for type have been dropped leading to the revised dataset of 10354 observations for further analysis.

In [493]:
gg.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 10354 entries, 0 to 10840
Data columns (total 13 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   App             10354 non-null  object 
 1   Category        10354 non-null  object 
 2   Rating          10354 non-null  float64
 3   Reviews         10354 non-null  float64
 4   Size            8829 non-null   float64
 5   Installs        10354 non-null  float64
 6   Type            10354 non-null  object 
 7   Price           10354 non-null  float64
 8   Content Rating  10354 non-null  object 
 9   Genres          10354 non-null  object 
 10  Last Updated    10354 non-null  object 
 11  Current Ver     10346 non-null  object 
 12  Android Ver     10352 non-null  object 
dtypes: float64(5), object(8)
memory usage: 1.4+ MB
